arch/arc/core/mpu/arc_mpu_common_internal.h

```diff
@@ -0,0 +1,300 @@
+ */
+int arc_core_mpu_region(uint32_t index, uint32_t base, uint32_t size, uint32_t region_attr)
+{
+	if (index >= get_num_regions()) {
+		return -EINVAL;
+	}
+
+	region_attr &= AUX_MPU_RDP_ATTR_MASK;
+
+	_region_init(index, base, size, region_attr);
+
+	return 0;
+}
+
+#if defined(CONFIG_USERSPACE)
+
+/**
+ * @brief configure MPU regions for the memory partitions of the memory domain
+ *
+ * @param thread the thread which has memory domain
+ */
+void arc_core_mpu_configure_mem_domain(struct k_thread *thread)
+{
+	int region_index = get_region_index_by_type(THREAD_DOMAIN_PARTITION_REGION);
+	uint32_t num_partitions;
+	struct k_mem_partition *pparts;
+	struct k_mem_domain *mem_domain = NULL;
+
+	if (thread) {
+		mem_domain = thread->mem_domain_info.mem_domain;
+	}
+
+	if (mem_domain) {
+		LOG_DBG("configure domain: %p", mem_domain);
+		num_partitions = mem_domain->num_partitions;
+		pparts = mem_domain->partitions;
+	} else {
+		LOG_DBG("disable domain partition regions");
+		num_partitions = 0U;
+		pparts = NULL;
+	}
+
+	for (; region_index >= 0; region_index--) {
+		if (num_partitions) {
+			LOG_DBG("set region 0x%x 0x%lx 0x%x",
+				 region_index, pparts->start, pparts->size);
+			_region_init(region_index, pparts->start, pparts->size, pparts->attr);
+			num_partitions--;
+		} else {
+			/* clear the left mpu entries */
+			_region_init(region_index, 0, 0, 0);
+		}
+		pparts++;
+	}
+}
+
+/**
+ * @brief remove MPU regions for the memory partitions of the memory domain
+ *
+ * @param mem_domain the target memory domain
+ */
+void arc_core_mpu_remove_mem_domain(struct k_mem_domain *mem_domain)
+{
+	ARG_UNUSED(mem_domain);
+
+	int region_index = get_region_index_by_type(THREAD_DOMAIN_PARTITION_REGION);
+
+	for (; region_index >= 0; region_index--) {
+		_region_init(region_index, 0, 0, 0);
+	}
+}
+
+/**
+ * @brief reset MPU region for a single memory partition
+ *
+ * @param domain  the target memory domain
+ * @param partition_id  memory partition id
+ */
+void arc_core_mpu_remove_mem_partition(struct k_mem_domain *domain, uint32_t part_id)
+{
+	ARG_UNUSED(domain);
+
+	int region_index = get_region_index_by_type(THREAD_DOMAIN_PARTITION_REGION);
+
+	LOG_DBG("disable region 0x%x", region_index + part_id);
+	/* Disable region */
+	_region_init(region_index + part_id, 0, 0, 0);
+}
+
+/**
+ * @brief get the maximum number of free regions for memory domain partitions
+ */
+int arc_core_mpu_get_max_domain_partition_regions(void)
+{
+	return get_region_index_by_type(THREAD_DOMAIN_PARTITION_REGION) + 1;
+}
+
+/**
+ * @brief validate the given buffer is user accessible or not
+ */
+int arc_core_mpu_buffer_validate(void *addr, size_t size, int write)
+{
+	int r_index;
+
+	/*
+	 * For ARC MPU, smaller region number takes priority.
+	 * we can stop the iteration immediately once we find the
+	 * matched region that grants permission or denies access.
+	 *
+	 */
+	for (r_index = 0; r_index < get_num_regions(); r_index++) {
```

Comment:
As we use r_index only in loop probably it's better to define it in loop itself:
```
for (int r_index = 0; r_index < get_num_regions(); r_index++) {
```